posts/2020-02-12-using-dotty-union-types-with-akka-typed-part-II.adoc

```diff
@@ -0,0 +1,285 @@
+= Using Dotty Union types with Akka Typed - Part II
+eloots
+v1.0, 2020-02-19
+:title: Using Dotty Union types with Akka Typed - Part II
+:tags: [dotty, akka]
+
+== Introduction
+
+After posting the first article on https://www.lunatech.com/2020/02/using-dotty-union-types-with-akka-typed[Using Dotty Union types with Akka Typed], I received https://twitter.com/hseeberger/status/1227660145125199877?s=20[valuable feedback] via Twitter which made me decide to write this follow-up article.
+
+I will clean-up a few loose ends and add some new insights. Let's start by listing them:
+
+1. In the Scala 2 version of the code, a message adapter is used, but the `WrappedPongResponse` message that wraps the response is part of the `Pinger` actor's protocol.
+2. The Dotty solution suffers from the same problem. Here, the `Pinger` actor's protocol is extended with the `PingPong.Pong` message.
+3. In the fix for issue 2, we use a mechanism in the Akka 2.6.1 library (`narrow` method) and we'll look at how Dotty has the potential to simplify this approach.
+4. I'll finish up with summarising a few of the best practices for coding actors that are to a large extent based on the https://doc.akka.io/docs/akka/current/typed/style-guide.html#style-guide[Akka Typed Style guide].
+
+== Restoring the `Pinger` actor's protocol
+
+Let's have a look at the final version of the protocol of the `Pinger` actor using Scala 2.13 and Akka 2.6:
+
+[source,scala,linenums]
+----
+object Pinger {
+
+  sealed trait Command
+  case object SendPing extends Command
+  case object StopPingPong extends Command
+  final case class WrappedPongResponse(pong: PingPong.Response)
+    extends Command
+
+  ...
+}
+----
+
+The problem should be obvious: we shouldn't expose the `WrappedPongResponse` wrapper message as part of the `Pinger` actor's protocol. In other words, we have unnecessarily _widened_ the actor's protocol by adding the publicly accessible `WrappedPongResponse` message to it.
+
+Obviously, this is undesirable and the fix in this case is quite simple: we can mark this message as private and we're done:
+
+[source,scala,linenums]
+----
+object Pinger {
+
+  sealed trait Command
+  case object SendPing extends Command
+  case object StopPingPong extends Command
+  private final case class WrappedPongResponse(pong: PingPong.Response)
+    extends Command
+
+  ...
+}
+----
+
+== Restoring the `Pinger` actor's protocol - Dotty version
+
+The Dotty version of the `Pinger` actor suffers from the same problem. The protocol of the actor is _widened_ with the `PingPong.Pong` message. This means that _any_ actor can send a `Pong` message to the `Pinger` actor. The solution in this case is different from previous one:
+
+* The `apply` method in the `Pinger` object, will return a `Behavior[Command]` thus restoring the original protocol
+* The `Pinger` actor defines an internal behavior of type `Behavior[CommandIncludingResponses]`
```

Comment:
In the code snipped you have renamed `CommandsAndResponses` to `CommandsAndResponses`

Reply:
Good catch!